fleetspeak/src/server/grpcservice/client/client.py

```diff
@@ -143,13 +149,41 @@ def _KeepAliveLoop(self):
     except Exception as e:  # pylint: disable=broad-except
       logging.error("Exception in KeepAlive: %s", e)
 
-  def Send(self, message):
-    """Sends a message to the Fleetspeak server.
+  def _RetryLoop(self, func, timeout=None):
+    """Retries an operation until success or deadline.
+
+    Args:
+
+      func: The function to run. Must take a timeout, in seconds, as a single
+        parameter. If it raises grpc.RpcError and deadline has not be reached,
+        it will be run again.
+
+      timeout: Retries will continue until timeout seconds have passed.
+    """
+    if not timeout:
+      timeout = self.DEFAULT_TIMEOUT
+    deadline = time.time() + timeout
+    sleep = 1
+    while True:
+      try:
+        return func(timeout)
+      except grpcs.RpcError:
+        timeout = deadline - time.time()
+        if time.time() + sleep > timeout:
```

Comment:
I am kind of lost trying to follow the logic in this loop. `timeout` and `sleep` are durations whereas `time.time()` is a point (in time). So, `time.time() + sleep` is a point in time which we compare against some duration. It doesn't typecheck. And because we are in 2018 already, this condition should be almost always true until `timeout` value is enormous. Am I missing something?

Reply:
Good catch. I just copied from the loop that was there, but I think I wrote that one too.

Updated. Will have to check how well we are testing this.